### PR TITLE
OCPBUGS-37932: Always log AWS service endpoints

### PR DESCRIPTION
**Pull request by Miciah**
#### aws/dns: Move a to-do for elbv2 custom endpoint

Move a to-do comment closer to the relevant source code.

#### aws/dns: Remove a useless `break` statement

Remove a useless, misleading `break` statement.  Rewrite the relevant code to be clearer.  Add a comment to describe how the service endpoints configuration is interpreted.

The AWS DNS provider implementation loops over the configured service endpoints to set service endpoint for the various clients: route53, elb, and tagging.  Inside this loop is a `switch` statement.  The first `case` statement has a `break` statement.  It appears that the intent may have been to break the loop.  However, because it is inside the `switch` statement, it only breaks out of the switch.  As there is nothing between the `switch` and the bottom of the loop, the `break` itself has no effect and can be removed.

However, because this `case` statement is the first one in the switch, it can prevent subsequent `case` statements from executing, so although it doesn't break the loop, the `case` statement does make subsequent loop iterations no-ops.

This commit changes the logic to be simpler and easier to understand by using an `if` statement and making the `switch` statement simply match on the name of the service endpoint.

The existing `Test_NewProvider` test proves that this commit does not ultimately change the behavior.

#### aws/dns: Tidy up logging for custom endpoints

Fix capitalization, and use `fmt.Sprintf` as `log.Info` does not expand formats.

#### aws/dns: Always log AWS API endpoints

Log the endpoint for every AWS API that we use, irrespective of whether a custom endpoint was configured for that API.

#### aws/dns: Add test for `NewProvider`

Add a new test to verify that `NewProvider` uses the expected AWS service endpoints.

### CODE REVIEW

**Comment by Miciah** [comment]
This is not the output I was hoping for:
```text
2024-08-29T22:01:39.540Z	INFO	operator.dns	dns/controller.go:663	Creating resourcegroupstaggingapi client	{"endpoint": ""}
2024-08-29T22:01:39.540Z	INFO	operator.dns	dns/controller.go:663	Creating elb client	{"endpoint": ""}
2024-08-29T22:01:39.540Z	INFO	operator.dns	dns/controller.go:663	Creating elbv2 client	{"endpoint": ""}
2024-08-29T22:01:39.540Z	INFO	operator.dns	dns/controller.go:663	Creating route53 client	{"endpoint": ""}
```
from [the ingress operator logs](https://gcsweb-ci.apps.ci.l2s4.p1.openshiftapps.com/gcs/test-platform-results/pr-logs/pull/openshift_cluster-ingress-operator/1137/pull-ci-openshift-cluster-ingress-operator-master-e2e-aws-operator/1829254157333172224/artifacts/e2e-aws-operator/gather-extra/artifacts/pods/openshift-ingress-operator_ingress-operator-6f5ff759f6-n6dhf_ingress-operator.log) from [the last e2e-aws-operator job run](https://prow.ci.openshift.org/view/gs/test-platform-results/pr-logs/pull/openshift_cluster-ingress-operator/1137/pull-ci-openshift-cluster-ingress-operator-master-e2e-aws-operator/1829254157333172224).

**Comment by Miciah** [comment]
This is what I was hoping for, from [the ingress-operator pod logs](https://gcsweb-ci.apps.ci.l2s4.p1.openshiftapps.com/gcs/test-platform-results/pr-logs/pull/openshift_cluster-ingress-operator/1137/pull-ci-openshift-cluster-ingress-operator-master-e2e-aws-operator/1829375534627295232/artifacts/e2e-aws-operator/gather-extra/artifacts/pods/openshift-ingress-operator_ingress-operator-5c9974fd6c-hrp55_ingress-operator.log) from [the last e2e-aws-operator run](https://prow.ci.openshift.org/view/gs/test-platform-results/pr-logs/pull/openshift_cluster-ingress-operator/1137/pull-ci-openshift-cluster-ingress-operator-master-e2e-aws-operator/1829375534627295232):
```text
2024-08-30T06:06:15.123Z	INFO	operator.dns	dns/controller.go:663	Created tags client	{"endpoint": "https://tagging.us-east-1.amazonaws.com"}
2024-08-30T06:06:15.124Z	INFO	operator.dns	dns/controller.go:663	Created elb client	{"endpoint": "https://elasticloadbalancing.us-west-2.amazonaws.com"}
2024-08-30T06:06:15.124Z	INFO	operator.dns	dns/controller.go:663	Created elbv2 client	{"endpoint": "https://elasticloadbalancing.us-west-2.amazonaws.com"}
2024-08-30T06:06:15.124Z	INFO	operator.dns	dns/controller.go:663	Created route53 client	{"endpoint": "https://route53.amazonaws.com"}
```

**Comment by Miciah** [comment]
https://github.com/openshift/cluster-ingress-operator/compare/bdbc28a5fd66e99bc60d3375ed6172d63ee5263b..5ab00e7e0e63d0211d1c93740346aa9cd41ca874 moves the unit test to be the first commit so that it can show that the subsequent changes don't modify the behavior.

**Comment by Miciah** [comment]
https://github.com/openshift/cluster-ingress-operator/compare/5ab00e7e0e63d0211d1c93740346aa9cd41ca874..2d015b64dc1f2e701f6dfc4520390b6d6da8d2f0 improves the unit test and fixes some logic to preserve previous behavior when all 3 service endpoints are specified and then any of those 3 is specified again.

File: pkg/dns/aws/dns_test.go [hunk 1]
```diff
@@ -133,3 +133,86 @@ func Test_zoneIDFromResource(t *testing.T) {
...
+	for _, tc := range cases {
+		t.Run(tc.name, func(t *testing.T) {
+			validateServiceEndpointsFn = func(provider *Provider) error {
+				return nil
+			}
+			provider, err := NewProvider(tc.config, "0.0.0-0")
+			if !assert.NoError(t, err) {
+				return
+			}
+			assert.NotNil(t, provider.route53)
```
Comment (alebedev87) [inline pkg/dns/aws/dns_test.go:211]: Why only route53 can potentially be `nil`? Do we need to check other clients?

**Review by alebedev87 — COMMENTED** [review]
LGMT, just a question.

File: pkg/dns/aws/dns_test.go (hunk 1 above)
Comment (Miciah) [reply pkg/dns/aws/dns_test.go]: Only the tagging client can be nil.  https://github.com/openshift/cluster-ingress-operator/compare/5ba53aefdbcd097c9fd4c58ac556571c627a169c..8afaffbf8ddbe65565bad52eea6267b615eceec2 adds additional assertions.

**Comment by Miciah** [comment]
[ci/prow/e2e-aws-operator-techpreview failed](https://prow.ci.openshift.org/view/gs/test-platform-results/pr-logs/pull/openshift_cluster-ingress-operator/1137/pull-ci-openshift-cluster-ingress-operator-master-e2e-aws-operator-techpreview/1835946767246430208) because `TestRouterCompressionOperation` failed in its cleanup:
```text
 === RUN   TestAll/serial/TestRouterCompressionOperation
    router_compression_test.go:316: GET hello-pod-lb4h5-openshift-ingress.apps.ci-op-s10nyts2-9e7c5.origin-ci-int-aws.dev.rhcloud.com failed: status 503, expected 200, retrying...
    router_compression_test.go:316: GET hello-pod-lb4h5-openshift-ingress.apps.ci-op-s10nyts2-9e7c5.origin-ci-int-aws.dev.rhcloud.com failed: status 503, expected 200, retrying...
    router_compression_test.go:257: Failed to delete hello-pod-lb4h5: Delete "https://api.ci-op-s10nyts2-9e7c5.origin-ci-int-aws.dev.rhcloud.com:6443/apis/route.openshift.io/v1/namespaces/openshift-ingress/routes/hello-pod-lb4h5": read tcp 10.131.243.238:58100->13.59.99.70:6443: read: connection reset by peer 
```
This looks like an issue with the API server or NLB or some other networking issue.

**Pull request merged**